codefresh/resource_pipeline.go

```diff
@@ -694,17 +694,78 @@ func extractStagesAndSteps(originalYamlString string) (stages, steps string) {
 					}
 				}
 			}
-		}
-		if item.Key == "stages" {
+		case "stages":
 			// For Stages we don't have ordering issue because it's a list
 			y, _ := yaml.Marshal(item.Value)
 			j2, _ := ghodss.YAMLToJSON(y)
 			stages = string(j2)
+		case "hooks":
+			switch hooks := item.Value.(type) {
+			default:
+				log.Fatalf("unsupported value type: %T", item.Value)
+
+			case yaml.MapSlice:
+				numberOfHooks := len(hooks)
+				for indexHook, hook := range hooks {
+					// E.g. on_finish
+					hooksBuilder.WriteString("\"" + hook.Key.(string) + "\" : {")
+					numberOfAttributes := len(hook.Value.(yaml.MapSlice))
+					for indexAttribute, hookAttribute := range hook.Value.(yaml.MapSlice) {
+						attribute := hookAttribute.Key.(string)
+						switch attribute {
+						case "steps":
+							hooksBuilder.WriteString("\"steps\" : {")
+							numberOfSteps := len(hookAttribute.Value.(yaml.MapSlice))
+							for indexStep, step := range hookAttribute.Value.(yaml.MapSlice) {
+								// We only need to preserve order at the first level to guarantee order of the steps, hence the child nodes can be marshalled
+								// with the standard library
+								y, _ := yaml.Marshal(step.Value)
+								j2, _ := ghodss.YAMLToJSON(y)
+								hooksBuilder.WriteString("\"" + step.Key.(string) + "\" : " + string(j2))
+								if indexStep < numberOfSteps-1 {
+									hooksBuilder.WriteString(",")
+								}
+							}
+							hooksBuilder.WriteString("}")
+						default:
+							// For Other elements we don't need to preserve order
+							y, _ := yaml.Marshal(hookAttribute.Value)
+							j2, _ := ghodss.YAMLToJSON(y)
+							hooksBuilder.WriteString("\"" + hookAttribute.Key.(string) + "\" : " + string(j2))
+						}
+
+						if indexAttribute < numberOfAttributes-1 {
+							hooksBuilder.WriteString(",")
+						}
+					}
+					hooksBuilder.WriteString("}")
+					if indexHook < numberOfHooks-1 {
+						hooksBuilder.WriteString(",")
+					}
+				}
+			}
+		case "mode":
+			pipeline.Spec.Mode = item.Value.(string)
+		case "fail_fast":
+			pipeline.Spec.FailFast = item.Value.(bool)
```

Comment:
In case if "fail_fast" is false, it will not be propagated to the pipeline spec, because the JSON marshaller will interpret "false" as "empty" and omit it